hibernate-core/src/main/java/org/hibernate/query/sqm/internal/AbstractSqmSelectionQuery.java

```diff
@@ -73,8 +95,169 @@ protected void errorOrLogForPaginationWithCollectionFetch() {
 	}
 
 	public abstract SqmStatement<R> getSqmStatement();
+	protected abstract void setSqmStatement(SqmSelectStatement<R> statement);
 	public abstract DomainParameterXref getDomainParameterXref();
 	public abstract TupleMetadata getTupleMetadata();
+
+	private SqmSelectStatement<R> getSqmSelectStatement() {
+		final SqmStatement<R> sqmStatement = getSqmStatement();
+		if ( sqmStatement instanceof SqmSelectStatement ) {
+			return (SqmSelectStatement<R>) sqmStatement;
+		}
+		else {
+			throw new IllegalSelectQueryException( "Not a select query" );
+		}
+	}
+
+	@Override
+	public SelectionQuery<R> setOrder(List<Order<? super R>> orderList) {
+		SqmSelectStatement<R> sqm = getSqmSelectStatement();
+		sqm = sqm.copy( noParamCopyContext() );
+		final SqmSelectStatement<R> select = sqm;
+		sqm.orderBy( orderList.stream().map( order -> sortSpecification( select, order ) )
+				.collect( toList() ) );
+		// TODO: when the QueryInterpretationCache can handle caching criteria queries,
+		//       simply cache the new SQM as if it were a criteria query, and remove this:
+		getQueryOptions().setQueryPlanCachingEnabled( false );
+		setSqmStatement( sqm );
+		return this;
+	}
+
+
+	@Override
+	public SelectionQuery<R> setOrder(Order<? super R> order) {
+		SqmSelectStatement<R> sqm = getSqmSelectStatement();
+		sqm = sqm.copy( noParamCopyContext() );
+		sqm.orderBy( sortSpecification( sqm, order ) );
+		// TODO: when the QueryInterpretationCache can handle caching criteria queries,
+		//       simply cache the new SQM as if it were a criteria query, and remove this:
+		getQueryOptions().setQueryPlanCachingEnabled( false );
+		setSqmStatement( sqm );
+		return this;
+	}
+
+	@Override
+	public SelectionQuery<R> setPage(Page page) {
+		setMaxResults( page.getMaxResults() );
+		setFirstResult( page.getFirstResult() );
+		return this;
+	}
+
+	static class KeyedResult<R> {
+		final R result;
+		final List<Comparable<?>> key;
+		public KeyedResult(R result, List<Comparable<?>> key) {
+			this.result = result;
+			this.key = key;
+		}
+
+		public R getResult() {
+			return result;
+		}
+
+		public List<Comparable<?>> getKey() {
+			return key;
+		}
+	}
+
+	private SqmSelectStatement<KeyedResult<R>> keyed(List<Order<? super R>> keyDefinition, List<Comparable<?>> keyValues) {
+		@SuppressWarnings("unchecked")
+		final SqmSelectStatement<KeyedResult<R>> sqm =
+				(SqmSelectStatement<KeyedResult<R>>)
+						getSqmSelectStatement().copy( noParamCopyContext() );
+		final NodeBuilder builder = sqm.nodeBuilder();
+		final SqmQuerySpec<?> querySpec = sqm.getQuerySpec();
+		final SqmWhereClause whereClause = querySpec.getWhereClause();
+		final List<? extends JpaSelection<?>> items = querySpec.getSelectClause().getSelectionItems();
+		final List<SqmPath<?>> newItems = new ArrayList<>();
+		if ( items.size() == 1 ) {
+			final JpaSelection<?> selected = items.get(0);
+			for ( int i = 0; i < keyDefinition.size(); i++ ) {
```

Comment:
If there are more than one key definitions, the ordering is currently incorrect. Think about the keys `[1,1]` and `[1,2]` which are both unique, but share the same first element.
The correct semantics are implemented by HQL/SQL tuple comparison e.g. `(x,y) > (?1,?2)`, which are `x > ?1 or x = ?1 and y > ?2`. If any of the key definition expressions is nullable, the predicate becomes even more complicated as the correct ordering depends on scroll direction and null ordering of the respective key definition.

Reply:
Ah right, of course. Missed that.

Reply:
Actually, @beikov would it be better to just generate a tuple comparison instead of using `and`/`or`?

Reply:
Of course it's better, but you need the code you have right now anyway in case one of the ordering expressions is nullable.

Reply:
Yup, that's what I figured you were going to say.

So I'll leave it like that for now.